example/sparse_advection/sparse_advection_package.cpp

```diff
@@ -0,0 +1,242 @@
+
+namespace sparse_advection_package {
+using parthenon::UserHistoryOperation;
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+  auto pkg = std::make_shared<StateDescriptor>("sparse_advection_package");
+
+  Real cfl = pin->GetOrAddReal("SparseAdvection", "cfl", 0.45);
+  pkg->AddParam("cfl", cfl);
+  Real refine_tol = pin->GetOrAddReal("SparseAdvection", "refine_tol", 0.3);
+  pkg->AddParam("refine_tol", refine_tol);
+  Real derefine_tol = pin->GetOrAddReal("SparseAdvection", "derefine_tol", 0.03);
+  pkg->AddParam("derefine_tol", derefine_tol);
+
+  pkg->AddParam("init_size", static_cast<Real>(0.1));
+
+  // set starting positions
+  Real pos = 0.8;
+  pkg->AddParam("x0", RealArr_t{pos, -pos, -pos, pos});
+  pkg->AddParam("y0", RealArr_t{pos, pos, -pos, -pos});
+
+  // add velocities, field 0 moves in (-1,-1) direction, 1 in (1,-1), 2 in (1, 1), and 3
+  // in (-1,1)
+  Real speed = pin->GetOrAddReal("SparseAdvection", "speed", 1.0) / sqrt(2.0);
+  pkg->AddParam("vx", RealArr_t{-speed, speed, speed, -speed});
+  pkg->AddParam("vy", RealArr_t{-speed, -speed, speed, speed});
+  pkg->AddParam("vz", RealArr_t{0.0, 0.0, 0.0, 0.0});
+
+  // add sparse field
+  Metadata m({Metadata::Cell, Metadata::Independent, Metadata::WithFluxes,
+              Metadata::FillGhost, Metadata::Sparse},
+             std::vector<int>({1}));
+  parthenon::SparsePool pool("sparse", m);
+
+  for (int sid = 0; sid < NUM_FIELDS; ++sid) {
+    pool.Add(sid);
+  }
+  pkg->AddSparsePool(pool);
+
+  pkg->CheckRefinementBlock = CheckRefinement;
+  pkg->EstimateTimestepBlock = EstimateTimestepBlock;
+
+  return pkg;
+}
+
+AmrTag CheckRefinement(MeshBlockData<Real> *rc) {
+  // refine on advected, for example.  could also be a derived quantity
+  auto pmb = rc->GetBlockPointer();
+  auto pkg = pmb->packages.Get("sparse_advection_package");
+  std::vector<std::string> vars;
+  for (int var = 0; var < NUM_FIELDS; ++var) {
+    vars.push_back("sparse_" + std::to_string(var));
+  }
+  // type is parthenon::VariablePack<CellVariable<Real>>
+  const auto &v = rc->PackVariables(vars);
+
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire);
+
+  typename Kokkos::MinMax<Real>::value_type minmax;
+  pmb->par_reduce(
+      "advection check refinement", 0, v.GetDim(4) - 1, kb.s, kb.e, jb.s, jb.e, ib.s,
+      ib.e,
+      KOKKOS_LAMBDA(const int n, const int k, const int j, const int i,
+                    typename Kokkos::MinMax<Real>::value_type &lminmax) {
+        if (v.IsAllocated(n)) {
+          lminmax.min_val =
+              (v(n, k, j, i) < lminmax.min_val ? v(n, k, j, i) : lminmax.min_val);
+          lminmax.max_val =
+              (v(n, k, j, i) > lminmax.max_val ? v(n, k, j, i) : lminmax.max_val);
+        }
+      },
+      Kokkos::MinMax<Real>(minmax));
```

Comment:
Dimension 4 of a variable pack (and meshblock pack) is always the total number of components in the pack and the packs include allocated and unallocated fields. Which "sparse_*" are included in the pack depends on how it was constructed. There are many (really many) overloads of `PackVariables`. One can list specific sparse indices by name, or one can list sparse base names and then filter by specific sparse indices (which is useful is one wants to get all variables for a particular set of materials, for example). In this case, I've added all the sparse fields individually by full name (which is maybe a bad example, I'll change it to just using the one base name).

I think it makes sense to include unallocated fields in the packs, because otherwise we would have ragged meshblock packs, where the same variables would have different indices for different blocks. And that would make loops over meshblock packs really complicated. See here for an attempt at only having the loops go over allocated indices: https://github.com/lanl/parthenon/tree/jlippuner/allocated-indices Because the list of indices needs to be copied to the device for every loop, it's much slower than just checking `IsAllocated` in the loop on the device. The right solution is probably to explicitly have an outer and inner loop where the allocation status gets checked in the outer loop. Although I doubt this will actually be significantly more efficient than just checking `IsAllocated` in the inner loop.